src/evo/cbtx.cpp

```diff
@@ -465,6 +469,11 @@ std::optional<CCbTx> GetCoinbaseTx(const CBlockIndex* pindex)
 
 std::optional<std::pair<CBLSSignature, uint32_t>> GetNonNullCoinbaseChainlock(const CBlockIndex* pindex)
 {
+    // There's no CbTx before DIP0003 activation
+    if (DeploymentActiveAfter(pindex, Params().GetConsensus(), Consensus::DEPLOYMENT_DIP0003)) {
```

Comment:
Should be a part of `GetCoinbaseTx` (because that's where we get `CCbTx` from), should use `DeploymentActiveAt` (because pindex is the block we are trying to read, not the one after it), should be `if (_not_ active) return nullopt`. Should also drop all unrelated refactoring.

tldr: the patch should look like that

```diff
diff --git a/src/evo/cbtx.cpp b/src/evo/cbtx.cpp
index a5f8d6b1cd..48baccd9d9 100644
--- a/src/evo/cbtx.cpp
+++ b/src/evo/cbtx.cpp
@@ -454,6 +454,11 @@ std::optional<CCbTx> GetCoinbaseTx(const CBlockIndex* pindex)
         return std::nullopt;
     }
 
+    // There's no CbTx before DIP0003 activation
+    if (!DeploymentActiveAt(*pindex, Params().GetConsensus(), Consensus::DEPLOYMENT_DIP0003)) {
+        return std::nullopt;
+    }
+
     CBlock block;
     if (!ReadBlockFromDisk(block, pindex, Params().GetConsensus())) {
         return std::nullopt;
```

Reply:
Good catch for `DeploymentActiveAt`, I don't know why it worked locally o_O

I'd prefer to drop `GetCoinbaseTx` completely because it's a confusing that do a lot of checks inside and used in only one module without any benefits, see commits

Reply:
>I'd prefer to drop GetCoinbaseTx completely because it's a confusing that do a lot of checks inside and used in only one module without any benefits, see commits

I agree it's confusing but let's keep this PR focused